### ISSUE TITLE
kakoune daemon without forking?

### ISSUE THREAD

**Issue by theamazingwaffle** [mawww/kakoune#3618]
I have `kak -d -s "$KAKSESSION"` run on login, but when i log out, the forked process doesn't get killed, but i can't access the old session either. I have to `killall kak` because i don't see the process id kakoune gives me, i go straight into xorg. And i don't want to manually kill the demon anyway. I think a `--no-fork` flag would be nice so the users can decide what to do with the headless devil themselves.

```sh
# instead of
kak -d -s "$KAKSESSION"
# do
kak -d -s "$KAKSESSION" &
# now on logout the shell will kill the kak daemon, i think.
```

**Comment by johnisom** [comment]
A potential workaround for your situation is to use `kak -l` to list the active sessions, and for each one, execute `echo kill | kak -p <session_name>`. To kill all sessions you could use this:

``` bash
for session in $(kak -l); do
  echo kill | kak -p $session
done
```

And if you want to kill sessions that have modified & unsaved buffers, use this variant:

``` bash
for session in $(kak -l); do
  echo kill! | kak -p $session
done
```

**Comment by alexherbo2** [comment]
More interactively, you can do `kak -c "$KAKSESSION"` to reconnect to the session, then `kill` or `kill!`, like mentioned above.

**Comment by johnisom** [comment]
I realize that my comment is definitely not the answer you're looking for, and I do think a `--no-fork` option would be nice. Though depending on your specific situation, I think using `killall kak`, my proposed solution, or something similar would work nicely. 

To clarify, when you say "log in" and "log out", are you referring to logging in and out from your display manager? Was the code block you provided part of your `.xprofile`/`.xinitrc`? If so, then using `--no-fork` and running in the background with `&` does indeed make the most sense.

**Comment by theamazingwaffle** [comment]
>To clarify, when you say "log in" and "log out", are you referring to logging in and out from your display manager? Was the code block you provided part of your .xprofile/.xinitrc? If so, then using --no-fork and running in the background with & does indeed make the most sense.

The code is in my `.profile` which is executed when i log into the tty.

**Comment by johnisom** [comment]
> > To clarify, when you say "log in" and "log out", are you referring to logging in and out from your display manager? Was the code block you provided part of your .xprofile/.xinitrc? If so, then using --no-fork and running in the background with & does indeed make the most sense.
> 
> The code is in my `.profile` which is executed when i log into the tty.

This also may be unhelpful, but if you're using bash, you could execute `echo kill | kak -p <session_name>` in your `.bash_logout` file.

**Comment by johnisom** [comment]
To be clear, it's of my opinion that a `--no-fork` command line option would be a great feature to have.

**Comment by theamazingwaffle** [comment]
guys i think i can add a `--no-fork` myself, can you please jump-start me on where to make it not fork?

**Comment by johnisom** [comment]
I'm not too familiar with the codebase so I can't help out there. It looks like you managed to find it :+1:

**Comment by lenormf** [comment]
You should rather use a service manager like SystemD, rather than downgrade the daemonisation feature to cater your usecase.

**Comment by theamazingwaffle** [comment]
>downgrade  
It's an optional "downgrade". If systemd launches a process, i suppose it watches over it. Is systemd able to monitor the child process that appeared after forking? And even if systemd is able to, can other services do that too? Wouldn't it be _simpler_ to just run it without forking? I've already done it so it would be a shame if it was discarded for reasons other than technical ones.

**Comment by krobelus** [comment]
> I've already done it so it would be a shame if it was discarded for reasons other than technical ones.

That sounds strange. Every option comes with a cost - the interface becomes more cluttered.
It's a tradeoff where we need to think about how useful this option would be to the general user base.
That's a technical decision IMO.

A systemd service is one option if you want that behavior.

> when i log out, the forked process doesn't get killed, but i can't access the old session either.

How come you can't access the old session? That seems to work for me.

How are you launching Xorg? I'm sure there is a way to kill kak on logout. I don't really see the need (why not keep the session runnin), but for example if you use `.xinitrc`, then
`trap "echo kill | kak -p $KAKSESSION" EXIT` could work (probably even without the trap).

**Comment by theamazingwaffle** [comment]
IMO the default behaviour should be listening and you'd need a `-fork` flag for kak to fork it. Although really why does kakoune need to do something that other tools can do and manage? Isn't it trying to be, like, a better unix citizen or something? So why limit users in process management when you can just use tools such as the init system, some process manager or just bash to manage the processes? Hiding the daemon can be done by many other tools that can be more convenient to the user.

**Comment by caksoylar** [comment]
I might be missing something, but would there be a functional difference between a non-forking version of `kak -d -s "$KAKSESSION"` and `kak -s "$KAKSESSION" -ui dummy`?

**Comment by theamazingwaffle** [comment]
@caksoylar Oh well...

Ok so what do i do? Do I close the PR and the issue then?

**Comment by eraserhd** [comment]
I want to point out that, on Mac OS, launchd is the preferred method for managing services, and services managed by launchd cannot fork -- launchd handles this part, and while people have tried to write blocking wrappers for forking programs, they generally do not work.

When you use systemd, it is preferable for the service not to fork, although it has a work-around for this that allows systemd to track the forks (I think it's called execution contexts or something?)

The `-ui dummy` option creates a client as well as a server, except that the client is non-functional in some ways.  This causes some possible problems with plugins.  What if a plugin grabs a random client because it needs to tell Kakoune to open a new window, for example?

I think it would be nice to have a non-forking server option for launchd at the least.

**Comment by mawww** [comment]
Hello,

coming a bit late to the party,

I tend to agree, forking the server process in daemon mode is indeed a mistake, Kakoune should leave whatever external tool take care of demonizing it. I'd be tempted to just change the behaviour not to fork at all, and leave it to users to `setsid kak -d -s blah &` or just `kak -d -s blah` depending on what behaviour they actually want.

I wonder how much this would break user configuration, we need to make sure the fix is easy enough so that on next release we can just point to that in the startup message.

Regarding the dummy ui, I am not sure if it has any value anymore, IIRC I introduced it for unit testing, but it got superseeded by the json-ui thats actually testable.

Another thing is that Kakoune still would auto-fork a server in some cases, like if you suspend or exit the initial client while another client exists. (If you just run `kak` you get a single process, no split between client/server).

Feedback on changing -d to never fork ?

**Comment by occivink** [comment]
I'm in favor of never forking, it's simpler all around.

I also use the dummy UI for non-interactive text processing, like so 
```
kak -ui dummy -e '
try %{
    # do something
   quit! 0
} catch %{
   quit! 1
}'
```
A bit like `kak -f`, but more powerful. So I would like to keep it around.

**Comment by eraserhd** [comment]
Vote on this comment?, :+1: don't fork, :-1: fork, :confused: don't care.

**Comment by alexherbo2** [comment]
How to execute a command reliably after forking if we pass a `&`?

[`kak-shell`]

```
start() {
  # Start the session in daemon mode.
  setsid kak -s "$KAKOUNE_SESSION" -d < /dev/null > /dev/null 2>&1
  connect "$@"
}
```

[`kak-shell`]: https://github.com/alexherbo2/connect.kak/blob/master/bin/kak-shell

**Comment by eraserhd** [comment]
@alexherbo2 How do you mean?  `-E` should still work.  Or are you asking how to block until the socket is set up?

**Comment by theamazingwaffle** [comment]
hey, @mawww! Is anything decided? What do i need to do? Make another pull request that makes non-forking default?

**Comment by alexherbo2** [comment]
@eraserhd Block until the socket is set up.

**Comment by alexherbo2** [comment]
@mawww I have a race condition in [`kak-shell`] using `&`.

[`kak-shell`]: https://github.com/alexherbo2/connect.kak/blob/master/bin/kak-shell

**Comment by lenormf** [comment]
Maybe use a locking mechanism (for example `flock`) via `-E`?

**Comment by alexherbo2** [comment]
@lenormf Can you give me the snippet to fix [`kak-shell`]?

[`kak-shell`]: https://github.com/alexherbo2/connect.kak/blob/master/bin/kak-shell

**Comment by theowenyoung** [comment]
It seems that kak is still forking the process?

When I run:

```bash
kak -d -s dev &!
```
Then exit the terminal:

```bash
exit
```

Reopen the terminal

```bash
kak -l
```

`dev` is still there.

**Comment by Screwtapello** [comment]
I don't know what shell supports a `&!` operator (it's not dash, bash, or zsh) or exactly what it does, but generally if you use `&` then *you* are forking the process into the background.

Shells differ in whether they automatically close backgrounded jobs when the shell exits, but in bash:

- running `kak -d -s dev` and then closing the terminal will terminate the session abruptly (the socket is listed as "dev (dead)"
- running `kak -d -s dev &` and then closing the terminal will terminate the session abruptly
- running `kak -d -s dev &` and then cleanly exiting the shell with `exit` will daemonize the session

**Comment by theowenyoung** [comment]
Thanks @Screwtapello , I tried again, and it is.

Because I'm using tmux, I got the wrong situation.

So when I use tmux, I need to use `Ctrl-b &` or `Ctrl-b x` to kill window or pane.

`&!` is a zsh operator, which means zsh can directly exit the window instead prompt a confirmation message.

**Comment by Screwtapello** [comment]
Ah, right. I searched `zsh(1)` for `&!` but I forgot that's just a high-level summary that points to other manpages, I eventually found `&!` documented in `zshmisc(1)`.

So here's one more case I didn't describe above:

- running `kak -d -s dev &` then running `disown` to tell the shell to forget about the job, then closing the terminal, will daemonize the session

I believe zsh's `&!` is shorthand for `& disown`.
